knowledge/knowledge-contracts
# Proxy contract

## Testing Notes
- Run `npx truffle console --network ropsten`
- Create the contracts
```solidity
KnowledgeBaseLegacy.new()
legacy = KnowledgeProxy.at('address from last response')

KnowledgeProxy.new()
proxy = KnowledgeProxy.at('address from last response')

Payable.new()
proxy.upgradeTo('address from last response')

knowledge = Payable.at(proxy.address)
```
- Make a transfer in the legacy contract and check the balance
```
legacy.transfer('0x39bce192859a27e23d55b8d451f98ce7ce092d22', 10000000000)
legacy.balanceOf('0x39bce192859a27e23d55b8d451f98ce7ce092d22')
```
- Upgrade from legacy to new proxy system and check if the balance is still there
```solidity
legacy.setNextContract(knowledge.address)
knowledge.setPrevContract(legacy.address)

legacy.upgrade()
knowledge.balanceOf('0x39bce192859a27e23d55b8d451f98ce7ce092d22')
```
- Make another transfer to test the proxy upgrades
```solidity
knowledge.transfer('0x39bce192859a27e23d55b8d451f98ce7ce092d22', 10000000000)
knowledge.balanceOf('0x39bce192859a27e23d55b8d451f98ce7ce092d22')
```
- Create a new contract and upgrade to it
```
Payable.new()
proxy.upgradeTo('address from last response')
```
- Test the balance is still there
```solidity
knowledge.balanceOf('0x39bce192859a27e23d55b8d451f98ce7ce092d22')
```
- Double check the balances and transactions history using `https://ropsten.etherscan.io/token/{proxy address}`
